src/DssDirectDepositHub.sol

```diff
@@ -0,0 +1,465 @@
+        daiJoin.join(address(this), total);
+
+        // normalized debt == erc20 DAI to pool (Vat rate for this ilk fixed to 1 RAY)
+
+        if (mode == Mode.NORMAL) {
+            vat.frob(ilk, address(pool), address(pool), address(this), -int256(amount), -int256(amount));
+            vat.slip(ilk, address(pool), -int256(amount));
+            vat.move(address(this), vow, _mul(fees, RAY));
+        } else if (mode == Mode.MODULE_CULLED) {
+            vat.slip(ilk, address(pool), -int256(amount));
+            vat.move(address(this), vow, _mul(total, RAY));
+        } else {
+            // This can be done with the assumption that the price of 1 aDai equals 1 DAI.
+            // That way we know that the prev End.skim call kept its gap[ilk] emptied as the CDP was always collateralized.
+            // Otherwise we couldn't just simply take away the collateral from the End module as the next line will be doing.
+            vat.slip(ilk, address(end_), -int256(amount));
+            vat.move(address(this), vow, _mul(total, RAY));
+        }
+
+        emit Unwind(ilk, amount);
+    }
+
+    function exec(bytes32 ilk_) external {
+        D3MPoolLike pool = ilks[ilk_].pool;
+
+        uint256 availableAssets = pool.maxWithdraw();
+
+        if (vat.live() == 0) {
+            uint256 culled = ilks[ilk_].culled;
+            // MCD caged
+            require(end.debt() == 0, "DssDirectDepositHub/end-debt-already-set");
+            require(culled == 0, "DssDirectDepositHub/module-has-to-be-unculled-first");
+            _unwind(
+                ilk_,
+                pool,
+                type(uint256).max,
+                availableAssets,
+                Mode.MCD_CAGED
+            );
+        } else if (live == 0) {
+            uint256 culled = ilks[ilk_].culled;
+            // This module caged
+            _unwind(
+                ilk_,
+                pool,
+                type(uint256).max,
+                availableAssets,
+                culled == 1
+                ? Mode.MODULE_CULLED
+                : Mode.NORMAL
+            );
+        } else {
+            // Normal path
+            (uint256 totalAssets, uint256 targetAssets) = ilks[ilk_].plan.calcSupplies(availableAssets);
+
+            if (targetAssets > totalAssets) {
+                _wind(ilk_, pool, targetAssets - totalAssets);
+            } else if (targetAssets < totalAssets) {
+                _unwind(
+                    ilk_,
+                    pool,
+                    totalAssets - targetAssets,
+                    availableAssets,
+                    Mode.NORMAL
+                );
+            }
+        }
+    }
+
+    // --- Collect Interest ---
+    function reap(bytes32 ilk_) external {
+        D3MPoolLike pool = ilks[ilk_].pool;
+
+        require(vat.live() == 1, "DssDirectDepositHub/no-reap-during-shutdown");
+        require(live == 1, "DssDirectDepositHub/no-reap-during-cage");
+
+        uint256 assetBalance = pool.assetBalance();
```

Comment:
I'm wondering if we need to delegate interest calculation to the pool itself (i.e. `pool.accruedInterest()`) since I could envision it being calculated differently for difference external systems @talbaneth or @hexonaut any thoughts on this?

Reply:
So here indeed there is a different behaviour betweem aave and compound.
In aave assetBalance() would call adai.balanceOf() which take accrued interest into account (by looking at the current timestamp).
In compound there is no such `view` function and in order to get the up to date balance we should call the non view `balanceOfUnderlying`.
I suggest to make `assetBalance()` a non view function so we are aligned to both.

Reply:
BTW same issue exists for the view function `convertToShares()`, but I'm not sure we should change anything there as it is called only after a deposit operation (which at least at compound accrues interest internally).

Reply:
I think we can still conform to the API. I believe you can override `view` modifiers by using assembly. I opened a thread on Discord to discuss. Hunting down the example I found.

Reply:
So we can use the stored value:

```
function assetBalance() external view override returns (uint256) {
        Exp memory exchangeRate = Exp({mantissa: pool.exchangeRateStored()});
        (MathError mErr, uint balance) = mulScalarTruncate(exchangeRate, pool.balanceOf(address(this)));
        require(mErr == MathError.NO_ERROR, "balance could not be calculated");
        return balance;
}
```

Reply:
Really would prefer not to break the `view` if possible as it tends to ripple downstream and is a PITA.

Reply:
@iamchrissmith as discussed we want to add a `pool.accrueIfNeeded()` before each time `pool.assetBalance();` is callsed.

Reply:
should we just call that inside `assetBalance()` rather than updating the hub?

Reply:
I think not, `assetBalance()` should probably stay view, and also we probably not want to hide that we're doing it.

Reply:
some of the calc changes slipped in, but this change is done in https://github.com/makerdao/dss-direct-deposit/pull/31/commits/abc39178654e8e4c9e76a0312fbbb1c86762eb36